pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1292,17 +1280,15 @@ void receiveIndividualMessagesFromBatch(MessageMetadata msgMetadata, int redeliv
                 if (possibleToDeadLetter != null) {
                     possibleToDeadLetter.add(message);
                 }
-                lock.readLock().lock();
-                try {
+                internalPinnedExecutor.execute(() -> {
                     if (peekPendingReceive() != null) {
                         notifyPendingReceivedCallback(message, null);
                     } else if (enqueueMessageAndCheckBatchReceive(message) && hasPendingBatchReceive()) {
                         notifyPendingBatchReceivedCallBack();
                     }
-                } finally {
-                    lock.readLock().unlock();
-                }
-                singleMessagePayload.release();
+                    singleMessagePayload.release();
+                    tryTriggerListener();
```

Comment:
I think we don't need move this into cycle or this execute block, it's better to keep as it used to be, because we cannot use `Listener` while calling `receive `or `receiveAsync`.

Reply:
The behavior here is consistent, the previous logic is
```
if(){
}else{
}
triggerListener
```
I just put the triggerListener into the if and else respectively.

If I don’t do this, triggerListener will execute in parallel with the if-else logic

Reply:
You put `triggerListener` into the [cycle](https://github.com/apache/pulsar/blob/c1517ffbe21f767a1147633f71e0bc0153b5a7db/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L1240)